toxcore/onion_client.c

```diff
@@ -1555,117 +1567,123 @@ static void do_friend(Onion_Client *onion_c, uint16_t friendnum)
         return;
     }
 
-    if (!onion_c->friends_list[friendnum].is_valid) {
+    Onion_Friend *o_friend = &onion_c->friends_list[friendnum];
+
+    if (!o_friend->is_valid) {
         return;
     }
 
-    unsigned int interval = ANNOUNCE_FRIEND;
+    uint32_t interval;
+    const uint64_t tm = mono_time_get(onion_c->mono_time);
+    const bool friend_is_new = o_friend->run_count <= ANNOUNCE_FRIEND_RUN_COUNT_BEGINNING;
 
-    if (onion_c->friends_list[friendnum].run_count < RUN_COUNT_FRIEND_ANNOUNCE_BEGINNING) {
-        interval = ANNOUNCE_FRIEND_BEGINNING;
-    } else {
-        if (onion_c->friends_list[friendnum].last_seen == 0) {
-            onion_c->friends_list[friendnum].last_seen = mono_time_get(onion_c->mono_time);
+    if (!friend_is_new) {
+        if (o_friend->last_seen == 0) {
+            o_friend->last_seen = tm;
         }
 
-        uint64_t backoff_interval = (mono_time_get(onion_c->mono_time) -
-                                     onion_c->friends_list[friendnum].last_seen)
-                                    / ONION_FRIEND_BACKOFF_FACTOR;
+        // how often we ping a node for a friend depends on how many times we've already tried.
+        // the interval increases exponentially, as the longer a friend has been offline, the less
+        // likely the case is that they're online and failed to find us
+        const double c = pow(2.0, min_u16(MAX_RUN_COUNT_EXPONENT, (double)(o_friend->run_count - 2)));
```

Comment:
The exponent is always an integer. You can use `1 << exponent` instead of `pow`.